Mage.Sets/src/mage/cards/c/CallForAid.java

```diff
@@ -0,0 +1,165 @@
+            return false;
+        }
+
+        FilterPermanent filter = new FilterCreaturePermanent();
+        filter.add(new ControllerIdPredicate(opponent.getId()));
+        List<Permanent> opponentCreatures = game.getBattlefield().getActivePermanents(filter, source.getControllerId(), game);
+        filter = new FilterPermanent();
+        filter.add(new PermanentReferenceInCollectionPredicate(opponentCreatures, game));
+
+        //"Gain control of all creatures target opponent controls until end of turn"
+        new GainControlAllEffect(Duration.EndOfTurn, filter).apply(game, source);
+        game.getState().processAction(game);
+
+        //"Untap those creatures".
+        new UntapAllEffect(filter).apply(game, source);
+
+        //"They gain haste until end of turn"
+        game.addEffect(new GainAbilityAllEffect(HasteAbility.getInstance(), Duration.EndOfTurn, filter), source);
+
+        //"You can't attack that player this turn"
+        game.addEffect(new CallForAidCantAttackThatPlayerEffect().setTargetPointer(this.getTargetPointer().copy()), source);
+
+        //"You can't sacrifice those creatures this turn"
+        game.addEffect(new CallForAidYouCantSacrificeEffect(source.getControllerId(), filter), source);
+
+        return true;
+    }
+
+    @Override
+    public CallForAidEffect copy() {
+        return new CallForAidEffect(this);
+    }
+}
+
+class CallForAidCantAttackThatPlayerEffect extends RestrictionEffect {
+
+    CallForAidCantAttackThatPlayerEffect() {
+        super(Duration.EndOfTurn, Outcome.Detriment);
+    }
+
+    private CallForAidCantAttackThatPlayerEffect(final CallForAidCantAttackThatPlayerEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public CallForAidCantAttackThatPlayerEffect copy() {
+        return new CallForAidCantAttackThatPlayerEffect(this);
+    }
+
+    @Override
+    public boolean applies(Permanent permanent, Ability source, Game game) {
+        // Effect applies to all permanents (most likely creatures) controlled by the caster of "Call for Aid"
+        return Objects.equals(source.getControllerId(), permanent.getControllerId());
+    }
+
+    @Override
+    public boolean canAttack(Permanent attacker, UUID defenderId, Ability source, Game game, boolean canUseChooseDialogs) {
+        if (defenderId == null || attacker == null) {
+            return true;
+        }
+        // Should be only a single target in case of "Call for Aid"
+        return !this.getTargetPointer().getTargets(game,source).contains(defenderId);
+    }
+
+}
+
+class CallForAidYouCantSacrificeEffect extends ContinuousEffectImpl {
+
+    private final UUID playerId; // controller who is prevented from sacrificing
+    private final FilterPermanent filter; // filter containing permanents that can't be sacrificed
+
+    CallForAidYouCantSacrificeEffect(UUID playerId, FilterPermanent filter) {
+        super(Duration.WhileOnBattlefield, Layer.RulesEffects, SubLayer.NA, Outcome.Benefit);
```

Comment:
The card reads "You can't sacrifice those creatures this turn" so shouldn't `Duration.EndOfTurn` be used instead? The current implementation prevents you from sacrificing affected creatures the next turn if you would end up getting control of any of them again through other means like [[Act of Aggression]].

Reply:
You are definitely right, I just wrote that duration wrong